hub/core/chunk_engine.py

```diff
@@ -440,25 +536,36 @@ def read_sample_from_chunk(
     ) -> np.ndarray:
         """Read a sample from a chunk, converts the global index into a local index. Handles decompressing if applicable."""
 
-        expect_compressed = self.tensor_meta.sample_compression is not None
         dtype = self.tensor_meta.dtype
 
         enc = self.chunk_id_encoder
 
         buffer = chunk.memoryview_data
+
         local_sample_index = enc.translate_index_relative_to_chunks(global_sample_index)
         shape = chunk.shapes_encoder[local_sample_index]
-        sb, eb = chunk.byte_positions_encoder[local_sample_index]
-
-        buffer = buffer[sb:eb]
 
         if len(buffer) == 0:
             return np.zeros(shape, dtype=dtype)
 
-        if expect_compressed:
-            sample = decompress_array(buffer, shape)
-            if cast and sample.dtype != self.tensor_meta.dtype:
-                sample = sample.astype(self.tensor_meta.dtype)
+        chunk_compression = self.tensor_meta.chunk_compression
+        if chunk_compression:
+            if get_compression_type(chunk_compression) == "byte":
+                decompressed = chunk.decompressed_data(compression=chunk_compression)
+                sb, eb = chunk.byte_positions_encoder[local_sample_index]
+                return np.frombuffer(decompressed[sb:eb], dtype=dtype).reshape(shape)
+            else:
+                return chunk.decompressed_samples()[local_sample_index]
+        sb, eb = chunk.byte_positions_encoder[local_sample_index]
+        buffer = buffer[sb:eb]
+
+        sample_compression = self.tensor_meta.sample_compression
+        if sample_compression:
+            sample = decompress_array(
+                buffer, shape, dtype=dtype, compression=sample_compression
+            )
+            if cast and sample.dtype != dtype:
+                sample = sample.astype(dtype)
```

Comment:
i believe a lot of this code is duplicated? `_append_bytes_to_compressed_chunk`

Reply:
Thats append, this is read.